### PR TITLE
[Administration] Define Security Policy for password

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to define security policy for Add/Edit Contact, My Account, and manage password hash for upgrade.

A PR should also be done on gorgone to handle the SQL request with the new password table.

**Fixes** # MON-6465

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

<h2> How this pull request can be tested ? </h2>

- Create a User
- Define a password
- It should follow the default security policy or an error should be thrown
- Edit your own password in My Account
- It should be follow the default security policy or an error should be thrown

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -28,7 +27,8 @@
         "symfony/web-profiler-bundle": "^4.4",
         "twig/twig": "^2.0",
         "webmozart/assert": "^1.8",
-        "zircote/swagger-php": "^3.0"
+        "zircote/swagger-php": "^3.0",
+        "centreon/centreon-test-lib": "dev-MON-6465"
```
Comment (kduret) [inline composer.json:31]: when you will reset to `dev-master`, don't forget to reset alphabetical order too

Comment (jeremyjaouen) [reply composer.json:31]: indeed

File: features/bootstrap/ClapiContext.php [hunk 2]
```diff
@@ -90,4 +110,67 @@ public function theExportedFileIsSimilarToTheImportedFiled()
...
+        $exportFileLines = file(sys_get_temp_dir() . DIRECTORY_SEPARATOR . self::CONFIGURATION_EXPORT_FILENAME);
+        array_shift($exportFileLines);
+        array_pop($exportFileLines);
+        $clapiActions = [];
+        foreach ($exportFileLines as $line) {
+            $clapiCommand = explode(';', $line);
+            $clapiActions[] = $clapiCommand[0];
+        }
+        $clapiActions = array_merge(array_unique($clapiActions));
+        if (self::CLAPI_ACTIONS_ORDER != $clapiActions) {
```
Comment (kduret) [inline features/bootstrap/ClapiContext.php:151]:
```suggestion
        if (self::CLAPI_ACTIONS_ORDER !== $clapiActions) {
```

File: lib/Centreon/Object/Contact/Contact.php [hunk 3]
```diff
@@ -45,4 +45,125 @@ class Centreon_Object_Contact extends Centreon_Object
...
+    public function insert($params = [])
+    {
+        $sql = "INSERT INTO $this->table ";
+        $sqlFields = "";
+        $sqlValues = "";
+        $sqlParams = [];
+
+        // Store password value and remove it from the array to not inserting it in contact table.
+        if (isset($params['contact_passwd'])) {
+            $password = $params['contact_passwd'];
```
Comment (kduret) [inline lib/Centreon/Object/Contact/Contact.php:64]: hashed password will be managed in another ticket

File: lib/Centreon/Object/Contact/Contact.php [hunk 4]
```diff
@@ -45,4 +45,125 @@ class Centreon_Object_Contact extends Centreon_Object
...
+     *
+     * @param mixed $parameterNames
+     * @param int $count
+     * @param int $offset
+     * @param string $order
+     * @param string $sort
+     * @param array $filters
+     * @param string $filterType
+     * @return array
+     * @throws Exception
```
Comment (kduret) [inline lib/Centreon/Object/Contact/Contact.php:113]:
```suggestion
     * @inheritDoc
```

File: src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php [hunk 5]
```diff
@@ -365,7 +377,6 @@ private function createContact(array $contact): Contact
                 $page->setUrlOptions($contact['topology_url_opt']);
             }
         }
-
         return (new Contact())
```
Comment (kduret) [inline src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php:380]:
```suggestion

        return (new Contact())
```

File: www/class/centreon-clapi/centreonAPI.class.php [hunk 6]
```diff
@@ -501,38 +501,39 @@ public function checkUser($useSha1 = false, $isWorker = false)
...
-        $DBRESULT = $this->DB->query("SELECT *
-                 FROM contact
-                 WHERE contact_alias = '" . $this->login . "'
-                 AND contact_activate = '1'");
-
+        $DBRESULT = $this->DB->prepare(
+            "SELECT `contact`.*, `contact_password`.`password` AS `contact_passwd` FROM `contact`
+            INNER JOIN `contact_password` ON `contact_password`.`contact_id` = `contact`.`contact_id`
+            WHERE `contact_alias` = :contactAlias
+            AND `contact_activate` = '1' AND `contact_register` = '1' LIMIT 1"
```
Comment (kduret) [inline www/class/centreon-clapi/centreonAPI.class.php:508]: add order by password creation_date DESC

File: www/class/centreon-clapi/centreonAPI.class.php [hunk 7]
```diff
@@ -501,38 +501,39 @@ public function checkUser($useSha1 = false, $isWorker = false)
...
-        }
-
-        $DBRESULT = $this->DB->query("SELECT *
-                 FROM contact
-                 WHERE contact_alias = '" . $this->login . "'
-                 AND contact_activate = '1'");
-
+        $DBRESULT = $this->DB->prepare(
+            "SELECT `contact`.*, `contact_password`.`password` AS `contact_passwd` FROM `contact`
+            INNER JOIN `contact_password` ON `contact_password`.`contact_id` = `contact`.`contact_id`
```
Comment (kduret) [inline www/class/centreon-clapi/centreonAPI.class.php:506]:
```suggestion
            LEFT JOIN `contact_password` ON `contact_password`.`contact_id` = `contact`.`contact_id`
```

File: src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php [hunk 8]
```diff
@@ -146,8 +155,11 @@ public function findByAuthenticationToken(string $token): ?Contact
     {
         $statement = $this->db->prepare(
             $this->translateDbName(
-                "SELECT contact.*, t.topology_url, t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
+                "SELECT contact.*, cp.password AS contact_passwd, t.topology_url,
+                t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
                 FROM `:db`.contact
+                INNER JOIN `:db`.contact_password cp
```
Comment (kduret) [inline src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php:161]:
```suggestion
                LEFT JOIN `:db`.contact_password cp
```

File: src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php [hunk 9]
```diff
@@ -113,8 +119,11 @@ public function findByName(string $name): ?Contact
     public function findBySession(string $sessionId): ?Contact
     {
         $request = $this->translateDbName(
-            'SELECT contact.*, t.topology_url, t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
+            'SELECT contact.*, cp.password AS contact_passwd, t.topology_url,
+            t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
             FROM `:db`.contact
+            LEFT JOIN `:db`.contact_password cp
```
Comment (kduret) [inline src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php:125]: add order by password creation_date DESC

File: src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php [hunk 10]
```diff
@@ -83,8 +86,11 @@ public function findById(int $contactId): ?Contact
     public function findByName(string $name): ?Contact
     {
         $request = $this->translateDbName(
-            'SELECT contact.*, t.topology_url, t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
+            'SELECT contact.*, cp.password AS contact_passwd, t.topology_url,
+            t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
             FROM `:db`.contact
+            LEFT JOIN `:db`.contact_password cp
```
Comment (kduret) [inline src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php:92]: add order by password creation_date DESC

File: src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php [hunk 11]
```diff
@@ -54,13 +54,16 @@ public function __construct(DatabaseConnection $pdo)
     public function findById(int $contactId): ?Contact
     {
         $request = $this->translateDbName(
-            'SELECT contact.*, t.topology_url, t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
+            'SELECT contact.*, cp.password AS contact_passwd, t.topology_url,
+            t.topology_url_opt, t.is_react, t.topology_id, tz.timezone_name
             FROM `:db`.contact
+            LEFT JOIN `:db`.contact_password cp
```
Comment (kduret) [inline src/Centreon/Infrastructure/Contact/ContactRepositoryRDB.php:60]: add order by password creation_date DESC

File: www/class/centreonAuth.class.php [hunk 12]
```diff
@@ -288,11 +355,15 @@ protected function checkUser($username, $password, $token)
...
         if ($this->autologin == 0 || ($this->autologin && $token != "")) {
-            $dbResult = $this->pearDB->query(
-                "SELECT * FROM `contact` " .
-                "WHERE `contact_alias` = '" . $this->pearDB->escape($username, true) . "'" .
-                "AND `contact_activate` = '1' AND `contact_register` = '1' LIMIT 1"
+            $dbResult = $this->pearDB->prepare(
+                "SELECT `contact`.*, `contact_password`.`password` AS `contact_passwd` FROM `contact`
+                LEFT JOIN `contact_password` ON `contact_password`.`contact_id` = `contact`.`contact_id`
+                WHERE `contact_alias` = :contactAlias
+                AND `contact_activate` = '1' AND `contact_register` = '1' LIMIT 1
```
Comment (kduret) [inline www/class/centreonAuth.class.php:362]: add order by password creation_date DESC

File: www/include/Administration/myAccount/DB-Func.php [hunk 13]
```diff
@@ -201,3 +218,79 @@ function updateContact($contact_id = null)
     $centreon->user->email = $ret['contact_email'];
     $centreon->user->setToken(isset($ret['contact_autologin_key']) ? $ret['contact_autologin_key'] : "''");
 }
+
+function validatePasswordModification(array $fields)
```
Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:222]: add phpdoc

File: www/include/Administration/myAccount/DB-Func.php [hunk 14]
```diff
@@ -186,12 +175,40 @@ function updateContact($contact_id = null)
...
+        $statement = $pearDB->prepare(
+            'SELECT id, password, creation_date from `contact_password`
+            WHERE `contact_id` = :contactId ORDER BY `creation_date` DESC'
+        );
+        $statement->bindValue(':contactId', $contactId, PDO::PARAM_INT);
+        $statement->execute();
+        //If 3 or more passwords are saved, delete the oldest one.
+        if (($result = $statement->fetchAll()) && count($result) >= 3) {
+            $statement = $pearDB->prepare(
+                'DELETE FROM `contact_password` WHERE `creation_date` < :creationDate AND contact_id = :contactId'
```
Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:195]:
```suggestion
                'DELETE FROM `contact_password` WHERE `creation_date` <= :creationDate AND contact_id = :contactId'
```

Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:195]: filter in the array the minimum creation date, and delete it directly from id

File: www/include/Administration/myAccount/DB-Func.php [hunk 15]
```diff
@@ -201,3 +218,79 @@ function updateContact($contact_id = null)
...
+        }
+    };
+    if (strlen($password) < (int) $passwordPolicy['password_length']) {
+        $errors['contact_passwd'] = sprintf(
+            _("Your password should be %d characters long."),
+            (int) $passwordPolicy['password_length']
+        );
+    }
+    if ((bool) $passwordPolicy['uppercase_characters'] === true && !preg_match('/[A-Z]/', $password)) {
+        $errors['contact_passwd'] = _("Your password should contains uppercase characters.");
```
Comment (kduret) [inline www/include/Administration/myAccount/DB-Func.php:260]: manage translation

File: www/include/Administration/myAccount/formMyAccount.php [hunk 16]
```diff
@@ -57,6 +57,17 @@
...
 }
 
+/**
+ * Get the Security Policy for automatic generation password.
+ */
+try {
+    $statement = $pearDB->query("SELECT * from password_security_policy");
+} catch (\PDOException $e) {
+    return false;
+}
```
Comment (kduret) [inline www/include/Administration/myAccount/formMyAccount.php:67]:
```suggestion
$statement = $pearDB->query("SELECT * from password_security_policy");
```

File: www/include/Administration/myAccount/formMyAccount.php [hunk 17]
```diff
@@ -86,6 +97,8 @@
 
     // selected by default is Resources status page
     $cct['default_page'] = $cct['default_page'] ?: CentreonAuth::DEFAULT_PAGE;
+
+    $statement = $pearDB->prepare("SELECT ");
```
Comment (kduret) [inline www/include/Administration/myAccount/formMyAccount.php:101]:
```suggestion
```

File: www/include/Administration/parameters/ldap/ldap.php [hunk 18]
```diff
@@ -35,6 +35,7 @@
  */
 
 require_once _CENTREON_PATH_ . 'www/class/centreonLDAP.class.php';
+require_once _CENTREON_PATH_ . 'www/class/centreonLDAPAdmin.class.php';
```
Comment (kduret) [inline www/include/Administration/parameters/ldap/ldap.php:38]:
```suggestion
require_once _CENTREON_PATH_ . 'www/class/CentreonLDAPAdmin.class.php';
```

File: www/include/Administration/parameters/ldap/form.php [hunk 19]
```diff
@@ -39,6 +39,7 @@
 }
 
 require_once _CENTREON_PATH_ . 'www/class/centreonLDAP.class.php';
+require_once _CENTREON_PATH_ . 'www/class/centreonLDAPAdmin.class.php';
```
Comment (kduret) [inline www/include/Administration/parameters/ldap/form.php:42]:
```suggestion
require_once _CENTREON_PATH_ . 'www/class/CentreonLDAPAdmin.class.php';
```

File: www/include/common/javascript/keygen.js [hunk 20]
```diff
@@ -70,12 +71,13 @@ function generatePassword(what)
      * TODO (maybe): Rechange the type if typing into the box...
      */
     if (what == "aKey") {
-		document.getElementById('aKey').value = sPassword;
+		    document.getElementById('aKey').value = sPassword;
```
Comment (kduret) [inline www/include/common/javascript/keygen.js:74]:
```suggestion
            document.getElementById('aKey').value = sPassword;
```

File: www/include/configuration/configObject/contact/DB-Func.php [hunk 21]
```diff
@@ -616,29 +637,47 @@ function updateContact($contact_id = null)
...
+        $statement = $pearDB->prepare(
+            'SELECT id, password, creation_date from `contact_password`
+            WHERE `contact_id` = :contactId ORDER BY `creation_date` DESC'
+        );
+        $statement->bindValue(':contactId', $contactId, PDO::PARAM_INT);
+        $statement->execute();
+        //If 3 or more passwords are saved, delete the oldest one.
+        if (($result = $statement->fetchAll()) && count($result) >= 3) {
+            $statement = $pearDB->prepare(
+                'DELETE FROM `contact_password` WHERE `creation_date` < :creationDate AND contact_id = :contactId'
```
Comment (kduret) [inline www/include/configuration/configObject/contact/DB-Func.php:661]: same as above

File: www/include/configuration/configObject/contact/formContact.ihtml [hunk 22]
```diff
@@ -120,7 +120,7 @@
             </td>
         </tr>
         <tr class="list_one"><td class="FormRowField"><img class="helpTooltip" name="centreon_login"> {$form.contact_oreon.label}</td><td class="FormRowValue">{$form.contact_oreon.html}</td></tr>
-        {if $auth_type != 'ldap'}
+        {if $auth_type != 'ldap' && $o != "mc"}
```
Comment (kduret) [inline www/include/configuration/configObject/contact/formContact.ihtml:123]: manage formContactLight too

File: www/install/createTables.sql [hunk 23]
```diff
@@ -2427,10 +2426,23 @@ CREATE TABLE `password_security_policy` (
...
   `special_characters` enum('0', '1') NOT NULL DEFAULT '1',
-  `attempts` int(11) UNSIGNED NOT NULL DEFAULT 5,
-  `blocking_duration` int(11) UNSIGNED NOT NULL DEFAULT 900,
-  `password_expiration` int(11) UNSIGNED NOT NULL DEFAULT 7776000,
-  `delay_before_new_password` int(11) UNSIGNED NOT NULL DEFAULT 3600
+  `attempts` int(11) UNSIGNED DEFAULT 5,
+  `blocking_duration` int(11) UNSIGNED DEFAULT 900,
+  `password_expiration` int(11) UNSIGNED DEFAULT 7776000,
+  `delay_before_new_password` int(11) UNSIGNED DEFAULT 3600,
+  `can_reuse_password`enum('0', '1') NOT NULL DEFAULT '0'
```
Comment (kduret) [inline www/install/createTables.sql:2433]:
```suggestion
  `can_reuse_password` enum('0', '1') NOT NULL DEFAULT '0'
```

File: www/install/php/Update-22.04.0-beta.1.php [hunk 24]
```diff
@@ -28,30 +28,73 @@
...
+        $statement->bindValue(':password', $row['contact_passwd']);
+        $statement->bindValue(':contactId', $row['contact_id']);
+        $statement->bindValue(':creationDate', time());
+        $statement->execute();
+    }
+    $errorMessage = "Unable to drop column 'contact_passwd' from 'contact' table";
+    $dbResult = $pearDB->query("ALTER TABLE `contact` DROP COLUMN `contact_passwd`");
+    if ($pearDB->inTransaction()) {
+        $pearDB->commit();
     }
```
Comment (kduret) [inline www/install/php/Update-22.04.0-beta.1.php:87]:
```suggestion
```

File: www/install/php/Update-22.04.0-beta.1.php [hunk 25]
```diff
@@ -28,30 +28,73 @@
 
 try {
     $pearDB->beginTransaction();
```
Comment (kduret) [inline www/install/php/Update-22.04.0-beta.1.php:30]:
```suggestion
```

File: www/install/php/Update-22.04.0-beta.1.php [hunk 26]
```diff
@@ -28,30 +28,73 @@
...
+    $dbResult = $pearDB->query("ALTER TABLE `contact` DROP COLUMN `contact_passwd`");
+    if ($pearDB->inTransaction()) {
+        $pearDB->commit();
     }
-    $pearDB->commit();
 } catch (Exception $e) {
-    $pearDB->rollback();
+    if ($pearDB->inTransaction()) {
+        $pearDB->rollBack();
+    }
```
Comment (kduret) [inline www/install/php/Update-22.04.0-beta.1.php:91]:
```suggestion
```

File: www/class/centreonContact.class.php [hunk 27]
```diff
@@ -247,4 +248,99 @@ public function getObjectForSelect2($values = array(), $options = array())
...
+        $statement = $this->db->prepare(
+            'INSERT INTO `contact_password` (`password`, `contact_id`, `creation_date`)
+            VALUES (:password, :contactId, :creationDate)'
+        );
+        $statement->bindValue(':password', $hashedPassword, \PDO::PARAM_STR);
+        $statement->bindValue(':contactId', $contactId, \PDO::PARAM_INT);
+        $statement->bindValue(':creationDate', time(), \PDO::PARAM_INT);
+        $statement->execute();
+
+        $this->deleteOldPasswords($contactId);
```
Comment (kduret) [inline www/class/centreonContact.class.php:271]:
```suggestion
```

**Review by adr-mo — APPROVED** [review]

**Review by AssanDiomande — APPROVED** [review]

**Pull request merged**